crates/store/src/transaction.rs

```diff
@@ -477,8 +483,107 @@ impl StoreTransaction {
             &META_TIP_BLOCK_HASH_KEY,
             parent_block_hash.as_slice(),
         )?;
+        self.clear_block_state(&block.hash().into())?;
         Ok(())
     }
+
+    pub fn record_block_state(
+        &self,
+        block_hash: &H256,
+        tx_index: u32,
+        col: Col,
+        raw_key: &[u8],
+    ) -> Result<(), Error> {
+        let key = self.get_state_record_key(block_hash, tx_index, col)?;
+        self.insert_raw(COLUMN_BLOCK_STATE_RECORD, key.as_ref(), raw_key)
+    }
+
+    fn prune_block_state_record(&self, current_block_number: u64) -> Result<(), Error> {
+        if current_block_number <= NUMBER_OF_CONFIRMATION {
+            return Ok(());
+        }
+        let to_be_pruned_block_number = current_block_number - NUMBER_OF_CONFIRMATION - 1;
+        let block_hash = self.get_block_hash_by_number(to_be_pruned_block_number)?;
+        let block_hash = match block_hash {
+            Some(block_hash) => block_hash,
+            None => match to_be_pruned_block_number {
+                0u64 => return Ok(()),
+                _ => return Err(Error::from("Invalid block hash".to_owned())),
+            },
+        };
+        self.clear_block_state_record(&block_hash)
+    }
+
+    fn clear_block_state_record(&self, block_hash: &H256) -> Result<(), Error> {
+        self.clear_block_state_and_record(block_hash, false)
+    }
+
+    fn clear_block_state(&self, block_hash: &H256) -> Result<(), Error> {
+        self.clear_block_state_and_record(block_hash, true)
+    }
+
+    fn clear_block_state_and_record(
```

Comment:
Change this function to `iter_block_state_record() -> Iterator<...>`.
And implement the delete logic in the `clear_block_state` & `clear_block_state_record`